File: rust/herb-linter/src/rules/herb_disable_comment_no_duplicate_rules.rs
use std::collections::HashMap;

use crate::herb_disable::parse_herb_disable_content;

use herb::nodes::ERBContentNode;
use herb::Visitor;

rule_visitor!(NoDuplicateRulesVisitor);
define_parser_rule!(
  HerbDisableCommentNoDuplicateRulesRule,
  "herb-disable-comment-no-duplicate-rules",
  Warning,
  NoDuplicateRulesVisitor
);

impl Visitor for NoDuplicateRulesVisitor {
  fn visit_erb_content_node(&mut self, node: &ERBContentNode) {
    let opening_tag = match &node.tag_opening {
      Some(token) => &token.value,
      None => return,
    };

    if opening_tag != "<%#" {
      return;
    }

    let content = match &node.content {
      Some(token) => &token.value,
      None => return,
    };

    let herb_disable = match parse_herb_disable_content(content) {
      Some(hd) => hd,
      None => return,
    };

    let content_location = match &node.content {
      Some(token) => &token.location,
      None => return,
    };

    let mut seen: HashMap<String, usize> = HashMap::new();

    for (index, rule_detail) in herb_disable.rule_name_details.iter().enumerate() {
      if seen.contains_key(&rule_detail.name) {
        let location = rule_detail.location(content_location).unwrap_or_else(|| node.location.clone());

        self.add_offense(
          format!("Duplicate rule `{}` in `herb:disable` comment. Remove the duplicate.", rule_detail.name),
          location,
        );
      } else {
        seen.insert(rule_detail.name.clone(), index);
      }
    }
  }
}

File: javascript/packages/linter/src/rules/herb-disable-comment-no-duplicate-rules.ts
import { ParserRule } from "../types.js"
import { HerbDisableCommentParsedVisitor } from "./herb-disable-comment-base.js"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { ERBContentNode, ParseResult } from "@herb-tools/core"
import type { HerbDisableComment } from "../herb-disable-comment-utils.js"

class HerbDisableCommentNoDuplicateRulesVisitor extends HerbDisableCommentParsedVisitor {
  protected checkParsedHerbDisable(node: ERBContentNode, _content: string, herbDisable: HerbDisableComment): void {
    const seenRules = new Set<string>()

    const check = (name: string, offset: number, length: number) => {
      if (seenRules.has(name)) {
        const location = this.createRuleNameLocation(node, { name, offset, length })
        const message = `Duplicate rule \`${name}\` in \`herb:disable\` comment. Remove the duplicate.`
        this.addOffenseWithFallback(message, location, node)
        return
      }

      seenRules.add(name)
    }

    herbDisable.ruleNameDetails.forEach(ruleDetail => {
      check(ruleDetail.name, ruleDetail.offset, ruleDetail.length)
    })

    herbDisable.fileScopedEntries.forEach(entry => {
      check(entry.name, entry.nameOffset, entry.nameLength)
    })
  }
}

export class HerbDisableCommentNoDuplicateRulesRule extends ParserRule {
  static ruleName = "herb-disable-comment-no-duplicate-rules"
  static introducedIn = this.version("0.8.0")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "warning"
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new HerbDisableCommentNoDuplicateRulesVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}
